src/python/pants/rules/core/test.py

```diff
@@ -103,13 +114,23 @@ class AddressAndDebugRequest:
     address: Address
     request: TestDebugRequest
 
+@union
+@dataclass(frozen=True)
```

Comment:
Generally we don't want unions to be dataclasses or to have fields explicitly defined. Only the subclasses should be dataclasses. It doesn't make sense to be able to create a `CoverageDataBatch` - sort of like an abstract class.

Reply:
So, initially this was

```python
@union 
class CoverageDataBatch():
    pass
```

But mypy complained extensively about it being given more arguments than it should have on line 194:
```python
coverage_reports = await MultiGet(
            Get[CoverageReport](
                CoverageDataBatch, coverage_batch_cls(tuple(addresses_and_test_results))
            )
            for coverage_batch_cls, addresses_and_test_results in coverage_data_collections
        )
```

Would love to know how to write it as an abstract class while still passing mypy!

Reply:
This might do the trick:

```
@union
class CoverageDataBatch(ABC):
   @abstractmethod
   def __init__(addresses_and_test_results: Tuple[AddressAndTestResult, ...]) -> None:
      """Subclasses should accept this in their constructor."""
```